### PR TITLE
Contracts and Harnesses for `unchecked_neg`

### PR DESCRIPTION
**Pull request by Yenyun035**
Towards #59 

### Changes
* Added contracts for `unchecked_neg` (located in `library/core/src/num/int_macros.rs`)
* Added a harness for `unchecked_neg` of each signed integer type
  * `i8`, `i16`, `i32`, `i64`, `i128`, `isize` --- 6 harnesses in total.
* Fixed comments.

### Revalidation
1. Per the discussion in #59, we have to **build and run Kani from `feature/verify-rust-std` branch**.
2. To revalidate the verification results, run the following command. `<harness_to_run>` can be either `num::verify` to run all harnesses or `num::verify::<harness_name>` (e.g. `check_unchecked_neg_i8`) to run a specific harness.
```
kani verify-std  "path/to/library" \
    --harness <harness_to_run> \
    -Z unstable-options \
    -Z function-contracts \
    -Z mem-predicates
```

All default harnesses should pass.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/core/src/num/int_macros.rs [hunk 1]
```diff
@@ -1167,6 +1167,7 @@ macro_rules! int_impl {
         #[rustc_const_unstable(feature = "unchecked_neg", issue = "85122")]
         #[inline(always)]
         #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+        #[requires(!self.overflowing_neg().1)]
```
Comment (zhassan-aws) [inline library/core/src/num/int_macros.rs:1170]: This can be simplified to `self != Self::MIN`.

File: library/core/src/num/mod.rs [hunk 2]
```diff
@@ -1809,10 +1824,10 @@ mod verify {
     // `unchecked_sub` proofs
     //
     // Target types:
-    // i{8,16,32,64,128} and u{8,16,32,64,128} -- 12 types in total
+    // i{8,16,32,64,128,size} and u{8,16,32,64,128,size} -- 12 types in total
     //
     // Target contracts:
-    // #[requires(!self.overflowing_sub(rhs).1)] // Preconditions: No overflow should occur
```
Comment (zhassan-aws) [inline library/core/src/num/mod.rs:1815]: Why was this comment removed?

**Comment by Yenyun035** [comment]
@zhassan-aws Thank you for your reviews! I have fixed the precondition for `unchecked_neg` and fixed some comments.

File: library/core/src/num/int_macros.rs [hunk 3]
```diff
@@ -1167,6 +1167,7 @@ macro_rules! int_impl {
         #[rustc_const_unstable(feature = "unchecked_neg", issue = "85122")]
         #[inline(always)]
         #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+        #[requires(self != $SelfT::MIN)]
```
Comment (zhassan-aws) [inline library/core/src/num/int_macros.rs:1170]: Can we also add an `ensures` clause that checks that the result is what we expect it to be? Even though the focus in this challenge is on memory safety, but the `ensures` clause in this case should be simple enough to add as part of this PR.

**Review by zhassan-aws — APPROVED** [review]

File: library/core/src/num/int_macros.rs (hunk 3 above)
Comment (Yenyun035) [reply library/core/src/num/int_macros.rs:1170]: Of course. Does `ensures(self == -self)` make sense in this case?

Comment (zhassan-aws) [reply library/core/src/num/int_macros.rs:1170]: An `ensures` clause is in terms of a `result`, so I think the following should work:
```rust
#[ensures(|result| result == -self)]
```

Comment (Yenyun035) [reply library/core/src/num/int_macros.rs:1170]: Oops, yes. I did mean that, in terms of the result. Thanks for clarifying! I'll make the change now.

Comment (zhassan-aws) [reply library/core/src/num/int_macros.rs:1170]: Thanks. Good to go.

**Comment by Yenyun035** [comment]
@feliperodri @celinval @carolynzech May I request the second approval to merge this PR? Thank you very much :)

**Review by celinval — APPROVED** [review]
Thanks!

**Comment by Yenyun035** [comment]
@celinval @zhassan-aws Thank you! This PR can get merged now :)

**Pull request merged**